cmd/run.go

```diff
@@ -503,29 +505,62 @@ func init() {
 	runCmd.Flags().AddFlagSet(runCmdFlagSet())
 }
 
+func createFilesystems() map[string]afero.Fs {
+	// We want to eliminate disk access at runtime, so we set up a memory mapped cache that's
+	// written every time something is read from the real filesystem. This cache is then used for
+	// successive spawns to read from (they have no access to the real disk).
+	// Also initialize the same for `https` but the caching is handled manually in the loader package
+	osfs := afero.NewOsFs()
+	if runtime.GOOS == "windows" {
+		// This is done so that we can continue to use paths with /|"\" through the code but also to
+		// be easier to traverse the cachedFs later as it doesn't work very well if you have windows
+		// volumes
+		osfs = fsext.NewTrimFilePathSeparatorFs(osfs)
+	}
+	return map[string]afero.Fs{
+		"file":  fsext.NewCacheOnReadFs(osfs, afero.NewMemMapFs(), 0),
+		"https": afero.NewMemMapFs(),
+	}
+}
+
 // Reads a source file from any supported destination.
-func readSource(src, pwd string, fs afero.Fs, stdin io.Reader) (*lib.SourceData, error) {
+func readSource(src, pwd string, filesystems map[string]afero.Fs, stdin io.Reader) (*loader.SourceData, error) {
 	if src == "-" {
 		data, err := ioutil.ReadAll(stdin)
 		if err != nil {
 			return nil, err
 		}
-		return &lib.SourceData{Filename: "-", Data: data}, nil
+		return &loader.SourceData{URL: &url.URL{Path: "-", Scheme: "file"}, Data: data}, nil
 	}
-	abspath := filepath.Join(pwd, src)
-	if ok, _ := afero.Exists(fs, abspath); ok {
-		src = abspath
+	var srcLocalPath string
+	if filepath.IsAbs(src) {
+		srcLocalPath = src
+	} else {
+		srcLocalPath = filepath.Join(pwd, src)
+	}
+	srcLocalPath = filepath.Clean(afero.FilePathSeparator + srcLocalPath)
```

Comment:
hmm this seems strange if we're running on Windows - why is it ok to prepend the `FilePathSeparator` to an absolute path?

Reply:
Because we are stripping it in the `TrimFilePathSeparatorFs` ;)

Reply:
Ah, right... This whole thing feels like it's way, way more complicated than it needs to be... Unfortunately, I don't have a very clear idea how to simplify it... :disappointed: 

Case in point, I very strongly think that code in `cmd` should have absolutely no idea that `TrimFilePathSeparatorFs` exists, much less trying to accommodate it in such a way... :confused: To have such a seemingly low-level implementation detail exposed here is another sign that we need to rethink the abstractions... Can you at least add a comment here, so other people (or us, in the future) aren't confused by this like I was?

Reply:
Actually, why is this function and `createFilesystems()` here, instead of in `loader/`?

Reply:
No better reason than: nobody moved it and this is the only place where it used ... so ... 
IMO we(I) am going to refactor this at some point and possibly have an "object" that is basically the filesystems with methods such as Resolve and Load.
But given that it is only used here I am somewhat against moving it somewhere else just because this package should probably not have it ... at least for now.
I am adding some readSource tests though as I am sick of fixing and breaking things between windows and linux

Reply:
:+1: :+1: for tests of this code... it would probably be easier to test if it's not in `cmd` though :wink:  But even if you're leaving it here, please at least add a comment that explains this seemingly strange line...

Regarding any future refactoring, I think I sent you a proto-proposal on Slack? Basically, I think we should scrap the `loader` and just have an interface called `Environment` or an `ExecutionEnvironment` or something like that, with methods for `import`/`require()` and `open()`. And have 2-3 different implementations of that interface:
1) `Native` / `System` / `OS` / whatever - initialized once with the current working directory (so we don't pass it around everywhere, append slashes, etc. strange things we currently do :expressionless: ), and then used to read (and cache) files from the local file system and actual network.
2) `Archive` - initialized  from archives, never does any actual network connections or touches the real FS.
3) `Test` / `Dummy` - for simple tests, maybe just a `Native` environment with the OS FS swapped out with a in-memory one.

The rest of the code, especially in the higher-level parts in `cmd/` and `js/`, should have absolutely  no idea about any of the implementation details in these environments - Windows/Linux file slashes, remote HTTPS scripts or imports, etc. But all of that, or something better (since this probably isn't enough...) could wait until a future refactoring of the code...